### PR TITLE
Feature: server GET routes and API request tests

### PR DESCRIPTION
**Pull request by amolgajewar**

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+//app.use(express.static(__dirname + '' ));
+
+app.get('/api/alsovieweditems', (req, res) => {
+
+  var queryString = 'select * from alsovieweditems';
+  var queryArgs = [];
+
+  db.query(queryString, queryArgs, function(err, results) {
+    if (err) {
+      res.json(err);
```
Comment (NapoleonDLP) [inline server/app.js]: Is responding to the client with the actual error giving them too much info? Security concerns I believe.

File: server/app.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  });
+
+});
+
+app.get('/api/alsovieweditems/id/:id', (req, res) => {
+
+  var queryString = 'select * from alsovieweditems where id = ?';
+  var queryArgs = req.params.id.split(',');
+
+  db.query(queryString, queryArgs, function(err, results) {
```
Comment (NapoleonDLP) [inline server/app.js]: Same here. I don't think you need to send the client the actual error.

File: server/app.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+});
+
+app.get('/api/alsovieweditems/startid/:startId/endid/:endId', (req, res) => {
+
+  var queryString = 'select * from alsovieweditems where id >= ? and id <= ?';
+  var queryArgs = [req.params.startId, req.params.endId];
+
+  db.query(queryString, queryArgs, function(err, results) {
+    if (err) {
+      res.json(err);
```
Comment (NapoleonDLP) [inline server/app.js]: same

File: server/app.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+const bodyparser = require('body-parser');
+const db = require('./db/index.js');
+
+const app = express();
+
+app.set('port', 3004);
+
+app.use(bodyparser.json());
+
+app.use(bodyparser.urlencoded({extended: true}));
```
Comment (tf3) [inline server/app.js]: Why is this necessary?

File: server/app.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    } else {
+      res.status(200).json(results);
+    }
+  });
+
+});
+
+app.get('/api/alsovieweditems/id/:id', (req, res) => {
+
+  var queryString = 'select * from alsovieweditems where id = ?';
```
Comment (tf3) [inline server/app.js]: I see this causing problems for the integration of our services. This responds with just one item, not a list of items.

You do have a route below that responds with a list of items, but to use it you have to make a request with multiple ids. But where will those ids come from? The user is going to visit an item with one id, say 17. If we then make an API request for also-viewed items with id 17, we're going to get only a single item. But that's not what we want.

Comment (amolgajewar) [reply server/app.js]: I have implemented this route just for the sake of availability. It may not be necessary to use it for front end. I will keep this point in mind while implementing front end.

File: server/app.js (hunk 1 above)
Comment (amolgajewar) [reply server/app.js]: What error code you suggests for database error instead? I will also look into it.

File: server/app.js (hunk 4 above)
Comment (amolgajewar) [reply server/app.js]: Express 4 development has made Express an independent routing and middleware web framework while removing all built-in middleware from its core and hence Express versioning and releases are not affected by middleware updates.

Instead it asks user to add all the middleware as per one's requirement for ex. body-parser. 

app.use(bodyParser.json())is used to tell the system that you want json parser to be used.

bodyParser.urlencoded({extended: ...}) is used to tell the system whether you want to use a simple algorithm for shallow parsing (i.e. false) or complex algorithm for deep parsing that can deal with nested objects (i.e. true).

URL encoding and JSON encoding both allow to convert a (nested) object to string, but the format is different.

This is what I have come across from internet but you have any particular question you can specify.

Comment (tf3) [reply server/app.js]: But why do you need to use `bodyParser`? Your API just accepts GET requests. It doesn't need to deal with POST requests, which is what `bodyParser` helps with.

**Review by mmcphillips — APPROVED** [review]
This generally looks good to me. You look like you have good test coverage so far.

File: server/app.js (hunk 1 above)
Comment (NapoleonDLP) [reply server/app.js]: Maybe a 500 is good enough. Here is a cheat sheet with some explanation. https://www.cheatography.com/kstep/cheat-sheets/http-status-codes/

**Comment by amolgajewar** [comment]
Implemented code reviews suggested. Requesting review for the same.

**Review by NapoleonDLP — COMMENTED** [review]
Looks good to me.

File: seedAlsoViewed.js [hunk 6]
```diff
@@ -90,13 +93,14 @@ for (var i = 1; i < 101; i++) {
...
+  var categoryId = getCategoryId();
 
-  seedAlsoViewedItems.push([id, image, title, oldPrice, currentPrice, freeSheeping, shippingCost]);
+  seedAlsoViewedItems.push([id, image, title, oldPrice, currentPrice, freeSheeping, shippingCost, categoryId]);
 }
 
 
 /// insert seed data
-var queryString = 'insert into alsovieweditems (id, image, title, oldprice, currentprice, freeshipping, shippingcost) values ?';
+var queryString = 'insert into alsovieweditems (id, image, title, oldprice, currentprice, freeshipping, shippingcost, categoryid) values ?';
```
Comment (mmcphillips) [inline seedAlsoViewed.js]: Does this single question mark denote shorthand for a single insertion for each of your columns? I've only seen this written so far to appear like: 

`'insert into alsovieweditems (id, image, title, oldprice, currentprice, freeshipping, shippingcost, categoryid) values (?, ?, ?, ?, ?, ?, ?, ?)';`

File: server/app.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  var queryString = 'select categoryid from alsovieweditems';
+  var queryArgs = [];
+
+  db.query(queryString, queryArgs, function(err, results) {
+    if (err) {
+      console.log(err);
+      res.status(500).send();
+    } else {
+      var categoryids = [];
+      for (var i = 0; i < results.length; i++) {
```
Comment (mmcphillips) [inline server/app.js]: The style guide  opposes the use of the ++ incrementing of variables.

37 can be written like
 `for (let i = 0; i < results.length; i += 1) {`

File: test/server/app.test.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      var expectedResultLength = 10;
+      var actualResultLength = res.body.length;
+      expect(actualResultLength).toBe(expectedResultLength);
+
+      var expectedStartId = 15;
+      var actualStartId = res.body[0].id;
+      expect(actualStartId).toBe(expectedStartId);
+
+      var expectedEndId = 24;
+      var actualEndId = res.body[res.body.length - 1].id;
```
Comment (mmcphillips) [inline test/server/app.test.js]: These tests generally look good to me. 
Set up ESlint configured to the airbnb styleguide and remove all references of var.

**Review by mmcphillips — COMMENTED** [review]
Looks good, just give it a run through esLint for some minor changes.

File: test/server/app.test.js (hunk 8 above)
Comment (amolgajewar) [reply test/server/app.test.js]: Sure. Thanks.

File: seedAlsoViewed.js (hunk 6 above)
Comment (amolgajewar) [reply seedAlsoViewed.js]: You can see in my queryArgs I am sending an array of an array. I am doing multiple insertion query and my each element of queryArgs is an array with specified column values. Hence using single question mark to represent the same.
